### PR TITLE
#34 Adds functionality to display stem and leaf data representation.

### PR DESCRIPTION
**Pull request by alinashakya**
Please do the code review.

This pull request consists of functionality to
- get data points from the user
- generate a stem and leaf plot to represent the data provided by the user

### CODE REVIEW

File: src/com/lftechnology/java/training/alina/basic/Student.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	/**
+	 * Function used to display all the student performance details
+	 * 
+	 * @param students
+	 *            {@link ArrayList} arrays of student details
+	 * @param courses
+	 *            {@link ArrayList} arrays of course details
+	 * @author Alina Shakya <alinashakya@lftechnology.com>
+	 */
+	private static void displayStudentDetails(ArrayList<Student> students,
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/Student.java]: MAJOR Remove the unused method parameter(s) "courses". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1172)

File: src/com/lftechnology/java/training/alina/basic/Student.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+	static ArrayList<Student> students = new ArrayList<Student>();
+	static ArrayList<Course> courses = new ArrayList<Course>();
+
+	private static final Logger LOGGER = Logger.getLogger(Student.class.getName());
+	private String fullName;
+	private int standard;
+	private int rollNo;
+
+	public Student(float subject1, float subject2, float subject3,
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/Student.java]: MAJOR Constructor has 8 parameters, which is greater than 7 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00107)

File: src/com/lftechnology/java/training/alina/basic/Student.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+			LOGGER.log(Level.INFO, "Student lists : {0}", new Object[] { student });
+		}
+	}
+
+	public String toString() {
+		return "Name : " + this.fullName + ", Class : " + this.standard
+				+ ", Roll Number : " + this.rollNo + "," + super.toString();
+	}
+
+	public int compareTo(Course val) {
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/Student.java]: CRITICAL Override "equals(Object obj)" to comply with the contract of the "compareTo(T o)" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1210)

File: src/com/lftechnology/java/training/alina/basic/Course.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	/**
+	 * Function used to get total marks of subjects
+	 * 
+	 * @return totalMarks {@link Float} total marks in all subjects
+	 * @author Alina Shakya <alinashakya@lftechnology.com>
+	 */
+	public float getTotalMarks() {
+		return (this.subject1 + this.subject2 + this.subject3 + this.subject4 + this.subject5);
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/Course.java]: MAJOR Remove those useless parentheses. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: src/com/lftechnology/java/training/alina/basic/Course.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	/**
+	 * Function used to get percentage of a student
+	 * 
+	 * @return percentage {@link Float} percentage in particular standard
+	 * @author Alina Shakya <alinashakya@lftechnology.com>
+	 */
+	public float getPercentage() {
+		return (getTotalMarks() / 5);
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/Course.java]: MAJOR Remove those useless parentheses. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: src/com/lftechnology/java/training/alina/basic/Course.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+			grade = 'D';
+		} else if (percentage < 50 && percentage >= 40) {
+			grade = 'E';
+		} else {
+			grade = 'F';
+		}
+		return grade;
+	}
+
+	public String toString() {
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/Course.java]: MAJOR Add the "@Override" annotation above this method signature [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)

File: src/com/lftechnology/java/training/alina/basic/stemleaf/StemLeafRepresentation.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Scanner;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Java program to present data in stem and leaf form
+ * 
+ * @author Alina Shakya <alinashakya@lftechnology.com>
+ */
+public class StemLeafRepresentation {
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/stemleaf/StemLeafRepresentation.java:16]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1118)

File: src/com/lftechnology/java/training/alina/basic/stemleaf/StemLeafRepresentation.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+	private static final Logger LOGGER = Logger
+			.getLogger(StemLeafRepresentation.class.getName());
+	private static int dataPoints;
+	private static DataSet[] data;
+
+	public static void main(String[] args) {
+		try (Scanner scanner = new Scanner(System.in)) {
+			getDataPoints(scanner);
+			getStemLeaf();
+		} catch (InputMismatchException e) {
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/stemleaf/StemLeafRepresentation.java:29]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: src/com/lftechnology/java/training/alina/basic/CalculateGPA.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ * Write a program that calculates percentage and GPA of 5 students. User will
+ * enter the marks of each 5 students from the keyboard. Each of the students
+ * has following attributes: name, class, rollnum, and marks obtained in 5
+ * subjects. Calculate the GPA and show the student name and GPA by
+ * position.[higher GPA should display on top]
+ * 
+ * @author Alina Shakya <alinashakya@lftechnology.com>
+ */
+public class CalculateGPA {
+	private final static Logger LOGGER = Logger.getLogger(CalculateGPA.class.getName());
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/CalculateGPA.java]: MINOR Reorder the modifiers to comply with the Java Language Specification. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AModifiersOrderCheck)

File: src/com/lftechnology/java/training/alina/basic/CalculateGPA.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Write a program that calculates percentage and GPA of 5 students. User will
+ * enter the marks of each 5 students from the keyboard. Each of the students
+ * has following attributes: name, class, rollnum, and marks obtained in 5
+ * subjects. Calculate the GPA and show the student name and GPA by
+ * position.[higher GPA should display on top]
+ * 
+ * @author Alina Shakya <alinashakya@lftechnology.com>
+ */
+public class CalculateGPA {
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/CalculateGPA.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1118)

File: src/com/lftechnology/java/training/alina/basic/stemleaf/DataSet.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
+package com.lftechnology.java.training.alina.basic.stemleaf;
+
+/**
+ * Class DataSet consists of data points to display stem and leaf
+ * 
+ * @author Alina Shakya <alinashakya@lftechnology.com>
+ */
+public class DataSet implements Comparable<DataSet> {
+	public int dataSet;
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/stemleaf/DataSet.java]: MAJOR Rename field "dataSet" [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1700)
MAJOR Make dataSet a static final constant or non-public and provide accessors if needed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)

File: src/com/lftechnology/java/training/alina/basic/stemleaf/DataSet.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+	public int getDataSet() {
+		return dataSet;
+	}
+
+	public void setDataSet(int dataSet) {
+		this.dataSet = dataSet;
+	}
+
+	@Override
+	public int compareTo(DataSet data) {
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/stemleaf/DataSet.java:24]: CRITICAL Override "equals(Object obj)" to comply with the contract of the "compareTo(T o)" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1210)

File: src/com/lftechnology/java/training/alina/basic/stemleaf/DataSet.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+		return Integer.compare(this.dataSet, data.dataSet);
+	}
+
+	@Override
+	public String toString() {
+		return "" + dataSet;
```
Comment (bhuwang) [inline src/com/lftechnology/java/training/alina/basic/stemleaf/DataSet.java]: MAJOR Use "Integer.toString" instead. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2131)

**Comment by bhuwang** [comment]
SonarQube analysis reported 21 issues:
* CRITICAL 3 critical
* MAJOR 10 major
* MINOR 8 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/bhuwang/java_training/blob/aea866b2b0c3948256563c9c6dfc054b3a43d453/src/com/lftechnology/java/training/alina/basic/CalculateGPA.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00105)
* MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/bhuwang/java_training/blob/aea866b2b0c3948256563c9c6dfc054b3a43d453/src/com/lftechnology/java/training/alina/basic/Course.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00105)
* MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/bhuwang/java_training/blob/aea866b2b0c3948256563c9c6dfc054b3a43d453/src/com/lftechnology/java/training/alina/basic/MarksValidation.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00105)
* MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/bhuwang/java_training/blob/aea866b2b0c3948256563c9c6dfc054b3a43d453/src/com/lftechnology/java/training/alina/basic/Student.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00105)
* MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/bhuwang/java_training/blob/aea866b2b0c3948256563c9c6dfc054b3a43d453/src/com/lftechnology/java/training/alina/basic/stemleaf/DataSet.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00105)
* MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/bhuwang/java_training/blob/aea866b2b0c3948256563c9c6dfc054b3a43d453/src/com/lftechnology/java/training/alina/basic/stemleaf/NumberValidation.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00105)
* MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/bhuwang/java_training/blob/aea866b2b0c3948256563c9c6dfc054b3a43d453/src/com/lftechnology/java/training/alina/basic/stemleaf/StemLeafRepresentation.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00105)

**Comment by bhuwang** [comment]
SonarQube analysis reported 9 issues:
- CRITICAL 2 critical
- MAJOR 4 major
- MINOR 3 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
- MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/bhuwang/java_training/blob/ba2415b3c98866d456dda1e6ecaac31a8f3b62a8/src/com/lftechnology/java/training/alina/basic/stemleaf/DataSet.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00105)
- MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/bhuwang/java_training/blob/ba2415b3c98866d456dda1e6ecaac31a8f3b62a8/src/com/lftechnology/java/training/alina/basic/stemleaf/NumberValidation.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00105)
- MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/bhuwang/java_training/blob/ba2415b3c98866d456dda1e6ecaac31a8f3b62a8/src/com/lftechnology/java/training/alina/basic/stemleaf/StemLeafRepresentation.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00105)

File: src/com/lftechnology/java/training/alina/basic/MarksValidation.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+	 * @return {@link Boolean}
+	 * @author Alina Shakya <alinashakya@lftechnology.com>
+	 */
+	private boolean isValidNumber(int fieldValue, int minValue) {
+		if (fieldValue < minValue) {
+			LOGGER.log(Level.INFO, "Please enter number greater than 0 : ");
+			return false;
+		} else {
+			return true;
+		}
```
Comment (unitedbaral) [inline src/com/lftechnology/java/training/alina/basic/MarksValidation.java:56]: Better to use ternary operator here

**Pull request merged**